main/src/main/scala/sbt/Defaults.scala

```diff
@@ -214,8 +214,41 @@ object Defaults extends BuildCommon {
           VirtualTerminal.handler,
         ) ++ serverHandlers.value :+ ServerHandler.fallback
       },
-      uncachedStamper := Stamps.uncachedStamps(fileConverter.value),
-      reusableStamper := Stamps.timeWrapLibraryStamps(uncachedStamper.value, fileConverter.value),
+      timeWrappedStamper := Stamps
+        .timeWrapLibraryStamps(Stamps.uncachedStamps(fileConverter.value), fileConverter.value),
+      reusableStamper := {
+        val converter = fileConverter.value
+        val unmanagedCache = unmanagedFileStampCache.value
+        val managedCache = managedFileStampCache.value
+        val backing = timeWrappedStamper.value
+        new xsbti.compile.analysis.ReadStamps {
+          def getAllLibraryStamps()
+              : java.util.Map[xsbti.VirtualFileRef, xsbti.compile.analysis.Stamp] =
+            backing.getAllLibraryStamps()
+          def getAllProductStamps()
+              : java.util.Map[xsbti.VirtualFileRef, xsbti.compile.analysis.Stamp] =
+            backing.getAllProductStamps()
+          def getAllSourceStamps()
+              : java.util.Map[xsbti.VirtualFileRef, xsbti.compile.analysis.Stamp] =
+            new java.util.HashMap[xsbti.VirtualFileRef, xsbti.compile.analysis.Stamp]
+          def library(fr: xsbti.VirtualFileRef): xsbti.compile.analysis.Stamp = {
+            val path = converter.toPath(fr)
+            managedCache
+              .getOrElseUpdate(path, sbt.nio.FileStamper.Hash)
+              .map(_.stamp)
+              .getOrElse(backing.library(fr))
+          }
```

Comment:
Would this return farm hash or timestamp?

Reply:
It should be hash but because backing is time wrapped, it first computes the timestamp and only updates the hash if the timestamp has changed (assuming I understand correctly how `Stamps.timeWrapLibraryStamps` works).